[SYCL][FPGA] Add clang support for buffer_location property

MrSidims opened the pull request:
This is a compiler-time known accessor property which serves as an
optimization hint for a compiler on where exactly buffer was allocated.
This is needed when a board has multiple disjoint global memories that
must be managed explicitly by a programmer.

When the property is added as a template parameter of an accessor -
SemaSYCL will implicitly add an attribute to an OpenCL kernel parameters,
which were generated from SYCL kernel object. It is not allowed to use the
attribute explicitly in SYCL code.

When the attribute is applied, clang generates metadata attached to OpenCL
kernel. Number of values stored in the metadata is the same as number of kernel
parameters. Order of metadata values is following the order of pointer
kernel parameters. Metadata values are of an integer type and is being set
accordingly values passed through accessor property ``buffer_location``.
This values are mapped in hardware backend to the actual locations of buffers
(DDR, QDR etc). Default value passed in the metadata is '-1'.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

MrSidims commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -80,6 +80,10 @@ class Util {
   /// half class.
   static bool isSyclHalfType(const QualType &Ty);
 
+  /// Checks whether given clang type is a full specialization of the SYCL
```
*buffer_location

MrSidims commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1285,13 +1336,13 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     size_t ParamIndex = Params.size();
     for (const ParmVarDecl *Param : InitMethod->parameters())
-      addParam(BS, Param->getType().getCanonicalType());
+      addParam(BS, Param->getType().getCanonicalType(), 42);
```
To remove '42' , which was used for debugging.

MrSidims commented on clang/test/CodeGenSYCL/Inputs/sycl.hpp:
```diff
@@ -68,18 +68,25 @@ enum prop_type {
   context_bound,
   enable_profiling,
-  base_prop
+  base_prop,
```
To remove ,

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10982,6 +10982,8 @@ def warn_sycl_implicit_decl
     : Warning<"SYCL kernel cannot call a recursive function">,
       InGroup<SyclStrict>, DefaultError;
+def warn_sycl_implicit_attr_usage : Warning <
+  "%0 attribute cannot be used explicitly">, InGroup<IgnoredAttributes>;
```
What do you mean by this?  Why give this a spelling at all if you don't want users using it?

erichkeane commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -660,6 +660,16 @@ void CodeGenFunction::EmitOpenCLKernelMetadata(const FunctionDecl *FD,
+
+  if (const SYCLIntelBufferLocationAttr *A =
+          FD->getAttr<SYCLIntelBufferLocationAttr>()) {
+    std::vector<size_t> Args = A->getActualArgs();
```
There is no reason to copy this vector here.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1198,13 +1207,40 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+    const auto *AccTy =
+        cast<ClassTemplateSpecializationDecl>(FieldTy->getAsRecordDecl());
+
+    // TODO: when SYCL headers' part is ready - replace this 'if' with an assert
```
This cannot be an assert, since we don't necessarily control the headers.  This should likely be a diagnostic.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1198,13 +1207,40 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+    // is going to be for other compile time known accessor properties, hence
+    // it's not clear, how handle them in SemaSYCL. But in general property_list
+    // is a parameter pack and shall be handled appropriately.
+    const auto Prop = cast<TemplateArgument>(AccTy->getTemplateArgs()[5]);
```
This assumption of the location of the layout of the type needs to be diagnosed as well.  Probably in the field checker.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1198,13 +1207,40 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+    // it's not clear, how handle them in SemaSYCL. But in general property_list
+    // is a parameter pack and shall be handled appropriately.
+    const auto Prop = cast<TemplateArgument>(AccTy->getTemplateArgs()[5]);
+    QualType PropTy = Prop.getAsType();
```
As does this one.

erichkeane commented:
They'd likely need to be errors, and stop execution.  If the buffer isn't the right layout it seems that it should just be an ill-formed program.

MrSidims commented:
> They'd likely need to be errors, and stop execution. If the buffer isn't the right layout it seems that it should just be an ill-formed program.

I disagree.  As for now accessor properties are Intel extension and serves only for optimization purposes (semantically they have no affect, but metadata generation). Hence if some one wants to use Intel compiler and a third-party library I see no reasons to deny them this opportunity (emitting warning and stop whatever compiler was up to do with this property is good enough).

erichkeane commented:
> 
> 
> > They'd likely need to be errors, and stop execution. If the buffer isn't the right layout it seems that it should just be an ill-formed program.
> 
> I disagree. As for now accessor properties are Intel extension and serves only for optimization purposes (semantically they have no affect, but metadata generation). Hence if some one wants to use Intel compiler and a third-party library I see no reasons to deny them this opportunity (emitting warning and stop whatever compiler was up to do with this property is good enough).

In that case, you absolutely cannot make any assumptions on the layout of them.  The purpose of errors is to be an assertion type-thing, where we would require this buffer_location property to be a certain format.  If we don't emit those as errors, NO assumptions on them being valid can be made.  I'm alright with this, but it means you have to be much more careful with the implementation.

MrSidims commented:
> > > They'd likely need to be errors, and stop execution. If the buffer isn't the right layout it seems that it should just be an ill-formed program.
> > 
> > 
> > I disagree. As for now accessor properties are Intel extension and serves only for optimization purposes (semantically they have no affect, but metadata generation). Hence if some one wants to use Intel compiler and a third-party library I see no reasons to deny them this opportunity (emitting warning and stop whatever compiler was up to do with this property is good enough).
> 
> In that case, you absolutely cannot make any assumptions on the layout of them. The purpose of errors is to be an assertion type-thing, where we would require this buffer_location property to be a certain format. If we don't emit those as errors, NO assumptions on them being valid can be made. I'm alright with this, but it means you have to be much more careful with the implementation.

Sure, thanks! For now this implementation doesn't break any SYCL tests, even we have no support for this feature in the headers, so I assume I was quite careful :)

erichkeane commented:
> 
> 
> > > > They'd likely need to be errors, and stop execution. If the buffer isn't the right layout it seems that it should just be an ill-formed program.
> > > 
> > > 
> > > I disagree. As for now accessor properties are Intel extension and serves only for optimization purposes (semantically they have no affect, but metadata generation). Hence if some one wants to use Intel compiler and a third-party library I see no reasons to deny them this opportunity (emitting warning and stop whatever compiler was up to do with this property is good enough).
> > 
> > 
> > In that case, you absolutely cannot make any assumptions on the layout of them. The purpose of errors is to be an assertion type-thing, where we would require this buffer_location property to be a certain format. If we don't emit those as errors, NO assumptions on them being valid can be made. I'm alright with this, but it means you have to be much more careful with the implementation.
> 
> Sure, thanks! For now this implementation doesn't break any SYCL tests, even we have no support for this feature in the headers, so I assume I was quite careful :)

You weren't particularly, I see a number of assumptions you make, like that the template arguments are types, etc.  I pointed out a few, but everything you pull from the buffer_location property isn't something that can be counted on.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1284,14 +1334,22 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     size_t ParamIndex = Params.size();
-    for (const ParmVarDecl *Param : InitMethod->parameters())
-      addParam(BS, Param->getType().getCanonicalType());
+    auto ParamIt = InitMethod->parameters().begin();
+    if (*ParamIt) {
+      // Add meaningful argument (not '-1') to buffer_location attribute only
+      // for an accessor pointer
+      size_t BufferLocAttrArg = handleBufferLocationProperty(FieldTy);
+      addParam(BS, (*ParamIt)->getType().getCanonicalType(), BufferLocAttrArg);
+      ++ParamIt;
+      for (; ParamIt != InitMethod->parameters().end(); ++ParamIt)
+        addParam(BS, (*ParamIt)->getType().getCanonicalType(), -1);
+    }
```
Feels like I've seen this before. Can we unify it?

MrSidims replied:
Do you mean `handleAccessorType` and `handleSyclSpecialType` functions? :)

I have slightly rewrote this part (now I'm obtaining source location, and paths for FD and BS are different).

MrSidims commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1213,8 +1249,17 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
-      addParam(FD, Param->getType().getCanonicalType());
+    auto ParamIt = InitMethod->parameters().begin();
+    if (*ParamIt) {
+      // Add meaningful argument (not '-1') to buffer_location attribute only
+      // for an accessor pointer
```
Indeed, reworked by applying Erich's suggestion №3

Fznamznon replied:
How?

MrSidims replied:
Ah, I have probably misunderstood your comment. I thought you mean order of metadata values in general shouldn't depend on the order of init method parameters and applying Erich's comment (add the attribute not to the kernel itself, but to each ParmVarDecl) should help.

But you meant, that we can't rely on the order of the parameters generated for an accessor (aka pointer, int, int, int) and I shan't take the first parameter to add the attribute, correct?

If you insist - I can rework it in a way of checking if ParmVarDecl is of a pointer type and apply the attribute only for this case. But I think, that if the order in any time will change - the ABI will become broken, hence it's unlikely that it's going to happen. So may we just rely on the current behavior making the code a little bit efficient by having less checks?

Fznamznon replied:
> But you meant, that we can't rely on the order of the parameters generated for an accessor (aka pointer, int, int, int) and I shan't take the first parameter to add the attribute, correct?

Yep.

> If you insist - I can rework it in a way of checking if ParmVarDecl is of a pointer type and apply the attribute only for this case. But I think, that if the order in any time will change - the ABI will become broken, hence it's unlikely that it's going to happen. So may we just rely on the current behavior making the code a little bit efficient by having less checks?

Okay, imagine situation when someone will add a new accessor type and spend hours on debugging hard coded behavior.
Other situation - if someone will use this implementation of FE and other implementation of headers without secret knowledge that pointer must be the first argument.
Anyway, I haven't seen functions with thousands of arguments, so a couple of checks won't make the compiler less efficient :)

MrSidims replied:
Agreed, actually, with your comment applied, `handleSyclAccessorType` and `handleSyclSpecialType` started looking nice again, thanks!

MrSidims replied:
Done

MrSidims commented:
What have changed:
0. I figured out, that my previous implementation version is incorrect: instead of passing buffer_location property to accessor's template parameter, users will pass property_list there, so the implementation was adjusted. 
1. The attribute now is spellingless, also it's being applied to ParmVarDecl instead of Function declaration;
2. Lowered number of assumptions (or at least I think so), added several diagnostics. @erichkeane  some of these diagnostics remains untested, because current implementation of sycl.hpp disallows on CPP level some use cases, when they can fire. I can also add something like fake-sycl.hpp with other accessor/property_list possible implementations to test them, what do you think?

@Fznamznon spec for compiletime known properties can be found here: https://github.com/intel/llvm/pull/1925 . For buffer location itself only SPIR-V spec is public now: https://github.com/intel/llvm/pull/2084 , ping @mohammadfawaz and @GarveyJoe  to make SYCL spec public.

MrSidims commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1172,39 +1180,108 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+         Prop != TemplArg.pack_end(); ++Prop) {
+      QualType PropTy = Prop->getAsType();
+      if (Util::isSyclBufferLocationType(PropTy)) {
+        // If we have more than 1 buffer_location properties on a single
```
This one is untested, but can be tested with the current version of sycl.hpp, to fix

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -1994,6 +1994,24 @@ can be lowered.
+  let Category = DocCatFunction;
+  let Heading = "kernel_args_buffer_location";
+  let Content = [{
+The attribute is being implicitly applied to an OpenCL kernel parameters
```
This needs some serious word-smithing, but I'll give it a shot.  I'll let the others further clarify.

"This attribute is implicitly added to OpenCL kernel parameters generated from a SYCL kernel object.  It lacks a spelling, as it is not intended to be used by the programmer.

This attribute causes clang to generate metadata on the OpenCL kernel containing the number of kernel parameters.  The metadata contains an integer that is set according to the values passed through the ``accessor`` property ``buffer_location``.  These values are mapped to the actual locations of the buffers (such as DDR, QDR, etc).  The default value is -1.

MrSidims replied:
Thanks, I like it! I've also added 'pointer' to 'kernel parameters' in the first string.

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10969,6 +10969,12 @@ def warn_boolean_attribute_argument_is_not_valid: Warning<
+  "Can't apply %0 property twice to the same accessor">;
+def err_sycl_invalid_property_template_param : Error<
+  "%0 template parameter must be a "
+  "%select{parameter pack|type|compiletime known non-negative integer|"
```
```suggestion
  "%select{parameter pack|type|non-negative integer|"
```

erichkeane commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -1536,6 +1539,14 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
       // Get argument name.
       argNames.push_back(llvm::MDString::get(VMContext, parm->getName()));
+
+      auto *SYCLBufferLocationAttr =
```
Why doesn't 'no metadata' mean 'not set'?  Why is the -1 value necessary?

MrSidims replied:
I don't quiet get the question, but I think that I shall state it more explicitly: the idea is to add the attribute only for pointer kernel parameters, that were generated for accessors  annotated with the property, hence not every parameter has the attribute applied (even the accessor pointer can has no such attribute without the property set). '-1' here means, that compiler's backend shan't make assumptions on where exactly the buffer was allocated and shall follow a default protocol for load-store units resolution. 

And there is one more thing that is needed to be clarified. Number of metadata values equals number of (not only pointer) kernel parameters - it's an implementation detail. It historically happened that with such metadata we annotates every kernel parameter, so '-1' for non-pointer kernel parameters just exists and means nothing. I'll look through the code and find where such a comment fits the best.

erichkeane replied:
Hmm... I still find myself wondering, if the backend is intending to ignore the -1 metadata, why bother putting it there at all?  You say 'number of metadata values equals number of kernel parmeters', but is that necessarily true?

Basically my question is: Can we avoid making the -1 value and just replace it with 'no data' to mean the same thing?  It doesn't seem that there is a differentiation between the two.

mohammadfawaz replied:
The value -1 itself is not important. As long as we can figure out, from the metadata, that the buffer_location for that particular argument is empty, we're okay. For OpenCL, that used to be an empty string `!""`. But it is important to have the number of metadata values equal to the number of kernel parameters to avoid any ambiguity (keep the 1:1 mapping).

MrSidims replied:
Well, that was the original requirement to make '-1' to be the default value, I personally don't mind to change it (would need to make a change in the translator though).

mohammadfawaz replied:
I'm personally okay with -1 and would rather not have us change the translator now since that change is already in. We could always change that later I suppose if we think there's a better option.

erichkeane replied:
It seems like a waste to keep the -1s, but at least this isn't wasting space in the AST.  Resolve for now.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1284,14 +1372,23 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     size_t ParamIndex = Params.size();
-    for (const ParmVarDecl *Param : InitMethod->parameters())
-      addParam(BS, Param->getType().getCanonicalType());
+    auto ParamIt = InitMethod->parameters().begin();
```
Don't use auto here.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1213,8 +1290,19 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+    if (*ParamIt) {
+      // Add meaningful argument (not '-1') to buffer_location attribute only
+      // for an accessor pointer
+      size_t BufferLocAttrArg =
```
Why is setting -1 here necessary?  Why doesn't omitted' just mean that?

erichkeane replied:
Also, why size_t when the handleBufferLocationProperty returns int?  or vice versa?

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1172,39 +1180,108 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+
+  // Obtain an integer value stored in a template parameter of buffer_location
+  // property to pass it to buffer_location kernel attribute
+  int handleBufferLocationProperty(QualType FieldTy, SourceLocation Loc) {
```
It seems that this should just accept the accessor type, rather than the code at line 1219.

erichkeane replied:
Also, instead of doing this as handleBufferLocationProperty, it should probably be something like, "AddAttributeIfNecessary" and take the NewParam above.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1213,8 +1290,19 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+      size_t BufferLocAttrArg =
+          isAccessorType
+              ? handleBufferLocationProperty(FieldTy, FD->getLocation())
+              : -1;
```
size_t cannot be -1.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1172,39 +1180,108 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
 
     LastParamIndex = Params.size();
     Params.push_back(NewParam);
+    if (LocationID != -1)
```
With as much work happens to get this -1 to go through, I think I'd rather addParam return the NewParam here, and have us just add this afterwards if we determine it is necessary

MrSidims replied:
Done

MrSidims commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -1188,6 +1188,17 @@ def SYCLDeviceIndirectlyCallable : InheritableAttr {
 
+def SYCLIntelBufferLocation : InheritableAttr {
+  // No spelling, as this attribute can't be created in the source code.
+  let Spellings = [];
```
Seems like not, table gen errors out: Record "`SYCLIntelBufferLocation', field `Spellings' does not have a list initializer!"

erichkeane replied:
Huh, interesting.  Thanks for checking.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1284,14 +1372,23 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+      // Add meaningful argument (not '-1') to buffer_location attribute only
+      // for an accessor pointer
+      size_t BufferLocAttrArg =
+          handleBufferLocationProperty(FieldTy, BS.getBeginLoc());
```
BTW, I haven't seen any test cases with inheritance (this function handles inherited accessors).

MrSidims replied:
Added the tests, actually I wasn't sure why do we have `handleSyclAccessorType` function if we already have `handleSyclSpecialType`, so thanks for pointing on it out!

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -2820,6 +2917,23 @@ bool Util::isSyclSpecConstantType(const QualType &Ty) {
+bool Util::isSyclBufferLocationType(const QualType &Ty) {
+  const StringRef &Name = "buffer_location";
+  std::array<DeclContextDesc, 4> Scopes = {
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
+      // TODO: this doesn't belong to property namespace, instead it shall be
+      // in its own namespace. Change it, when the actual implementation in SYCL
+      // headers is ready
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "property"},
```
Extending this detecting approach is not really good, we already trying to avoid it https://github.com/intel/llvm/pull/2091 .

MrSidims replied:
I'm down with replacing this function when #2091 is ready.

Fznamznon replied:
How #2091 will help you to do it?

MrSidims replied:
I think, that at least `property_list` can be added as enum argument of this attribute and here we query for it, or I'm missing something important?

Fznamznon replied:
The idea is to make it generic, i.e. without arguments. See https://github.com/intel/llvm/pull/2091#discussion_r454241718
Although I'm not really sure how to make it work for specialization constants and any other classes that will need some handling that differs from handling for accessors.

MrSidims commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1172,39 +1180,108 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+      return -1;
+    const auto PropList = cast<TemplateArgument>(AccTy->getTemplateArgs()[5]);
+    if (PropList.getKind() != TemplateArgument::ArgKind::Type) {
+      SemaRef.Diag(Loc, diag::err_sycl_invalid_property_template_param)
+          << "accessor's 5th" << /*type*/ 1;
```
In this case a 3rd party library would crash without trails with our compiler, see @erichkeane comment from above.

erichkeane replied:
Yep!  That said, these checks should be in the field-checker.  None of the others are supposed to diagnose.

erichkeane replied:
Don't do these diagnostics in the DeclCreator, do them in the FieldChecker (above).  That way, we can then assume these things in this visitor.

MrSidims commented:
Addressed the comments, major:
- No more '-1' in SemaSYCL, yet I slightly lean towards having it as a default MD value;
- `handleBufferLocationProperty` was split into `handleAccessorPropertyList` and `handleBufferLocationProperty`

I left @Fznamznon comment https://github.com/intel/llvm/pull/2166#discussion_r462249604 unaddressed, since I don't really know, how to improve special type detection

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -1994,6 +1994,23 @@ can be lowered.
+according to the values passed through the ``accessor`` property
+``buffer_location``. These values are mapped to the actual locations of the
+global buffers (such as DDR, QDR, etc) and applied to pointer kernel parameters.
+The default value is -1.
```
Now that I see this, please clarify the last sentence here with the meaning/difference between -1 and no-metadata (and clarify that the default value for the metadata is -1)..  I realize it is pretty subtle, so I won't try to author it myself.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1200,11 +1209,82 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+    const auto PropList = cast<TemplateArgument>(AccTy->getTemplateArgs()[5]);
+    if (PropList.getKind() != TemplateArgument::ArgKind::Type) {
+      SemaRef.Diag(Loc, diag::err_sycl_invalid_property_template_param)
+          << "accessor's 5th" << /*type*/ 1;
```
Also, don't put things like "accessor's 5th" into the diagnostic.  This prevents internationalization/localization.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1200,11 +1209,82 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+        cast<ClassTemplateSpecializationDecl>(PropTy->getAsRecordDecl());
+    const auto BufferLoc = PropDecl->getTemplateArgs()[0];
+    if (BufferLoc.getKind() != TemplateArgument::ArgKind::Integral) {
+      SemaRef.Diag(Loc, diag::err_sycl_invalid_property_template_param)
```
Same here with the diagnostics.

MrSidims replied:
Every diagnostic, but a check of buffer_location duplication was move to FieldChecker

MrSidims commented:
Seems like CPU RT in our CI isn't ready for this change SPIR-V-wise, so I have reverted SPIR-V support of the feature for testing.

MrSidims commented on llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -559,15 +552,6 @@ SPIRVFunction *LLVMToSPIRV::transFunctionDecl(Function *F) {
-      // Order of integer numbers in MD node follows the order of function
-      // parameters on which we shall attach the appropriate decoration. Add
-      // decoration only if MD value is not negative.
-      BM->addCapability(CapabilityFPGABufferLocationINTEL);
```
Yeap, here is a bug, we shan't add the Capability, when we don't add the decoration. So I'd even say it's not CPU RT problem, but a bug in the translator part as well.

MrSidims commented:
This patch needs to be merged first: https://github.com/intel/llvm/pull/2240

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1076,6 +1084,51 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
+    }
+    for (TemplateArgument::pack_iterator Prop = TemplArg.pack_begin();
+         Prop != TemplArg.pack_end(); ++Prop) {
+      QualType PropTy = Prop->getAsType();
```
Pack elements don't have to be types.  This is potentially an assert and should likely be diagnosed as well.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1076,6 +1084,51 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
+
+  void checkBufferLocationType(QualType PropTy, SourceLocation Loc) {
+    const auto *PropDecl =
+        cast<ClassTemplateSpecializationDecl>(PropTy->getAsRecordDecl());
```
Even if it IS a type, the template parameter could be 'int'.  We also cannot make this assumption.

MrSidims replied:
But at the same time we have already checked `isSyclBufferLocationType` before calling `checkBufferLocationType`, or I'm missing something?

erichkeane replied:
Ah, right.  This one is probably OK then.  The rest aren't.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1076,6 +1084,51 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
+    }
+    const auto *PropListDecl =
+        cast<ClassTemplateSpecializationDecl>(PropListTy->getAsRecordDecl());
+    const auto TemplArg = PropListDecl->getTemplateArgs()[0];
```
Do we ever make sure that the template actually HAS arguments?  It could be a specialization without any.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1076,6 +1084,51 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
+  void checkBufferLocationType(QualType PropTy, SourceLocation Loc) {
+    const auto *PropDecl =
+        cast<ClassTemplateSpecializationDecl>(PropTy->getAsRecordDecl());
+    const auto BufferLoc = PropDecl->getTemplateArgs()[0];
```
Same problem with assuming template argument size.
